File: server/webapp/WEB-INF/rails.new/app/assets/javascripts/pipeline_configs/models/pipeline.js
/*
 * Copyright 2015 ThoughtWorks, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

define(['mithril', 'lodash', 'string-plus', './model_mixins', './environment_variables', './parameters', './materials', './tracking_tool', './stages'], function (m, _, s, Mixins, EnvironmentVariables, Parameters, Materials, TrackingTool, Stages) {
  var Pipeline = function (data) {
    this.constructor.modelType = 'pipeline';
    Mixins.HasUUID.call(this);

    this.name                  = m.prop(data.name);
    this.enablePipelineLocking = m.prop(data.enablePipelineLocking);
    this.templateName          = m.prop(s.defaultToIfBlank(data.templateName, ''));
    this.labelTemplate         = m.prop(s.defaultToIfBlank(data.labelTemplate, ''));
    this.timer                 = m.prop(s.defaultToIfBlank(data.timer, new Pipeline.Timer({})));
    this.environmentVariables  = s.overrideToJSON(m.prop(s.defaultToIfBlank(data.environmentVariables, new EnvironmentVariables())));
    this.parameters            = s.overrideToJSON(m.prop(s.defaultToIfBlank(data.parameters, new Parameters())));
    this.materials             = s.overrideToJSON(m.prop(s.defaultToIfBlank(data.materials, new Materials())));
    this.trackingTool          = s.overrideToJSON(m.prop(data.trackingTool));
    this.stages                = s.overrideToJSON(m.prop(s.defaultToIfBlank(data.stages, new Stages())));

    this.validate = function () {
      var errors = new Mixins.Errors();

      if (s.isBlank(this.labelTemplate())) {
        errors.add('labelTemplate', Mixins.ErrorMessages.mustBePresent('labelTemplate'));
      } else if (!this.labelTemplate().match(/(([a-zA-Z0-9_\-.!~*'()#:])*[$#]\{[a-zA-Z0-9_\-.!~*'()#:]+(\[:(\d+)])?}([a-zA-Z0-9_\-.!~*'()#:])*)+/)) {
        errors.add('labelTemplate', "Label should be composed of alphanumeric text, it may contain the build number as ${COUNT}, it may contain a material revision as ${<material-name>} or ${<material-name>[:<length>]}, or use params as #{<param-name>}");
      }

      return errors;
    };
  };

  Pipeline.get = function (url) {
    return m.request({method: 'GET', url: url});
  };

  Pipeline.fromJSON = function (data) {
    return new Pipeline({
      name:                  data.name,
      enablePipelineLocking: data.enable_pipeline_locking,
      templateName:          data.template_name,
      labelTemplate:         data.label_template,
      timer:                 Pipeline.Timer.fromJSON(data.timer),
      trackingTool:          TrackingTool.fromJSON(data.tracking_tool),
      environmentVariables:  EnvironmentVariables.fromJSON(data.environment_variables),
      parameters:            Parameters.fromJSON(data.parameters),
      materials:             Materials.fromJSON(data.materials),
      stages:                Stages.fromJSON(data.stages)
    });
  };

  Pipeline.Timer = function (data) {
    this.constructor.modelType = 'pipelineTimer';
    Mixins.HasUUID.call(this);

    this.spec          = m.prop(s.defaultToIfBlank(data.spec, ''));
    this.onlyOnChanges = m.prop(data.onlyOnChanges);
  };

  Pipeline.Timer.fromJSON = function (data) {
    if (!_.isEmpty(data)) {
      return new Pipeline.Timer({
        spec:          data.spec,
        onlyOnChanges: data.only_on_changes
      });
    }

  };

  return Pipeline;
});

File: domain/src/main/java/com/thoughtworks/go/domain/Pipeline.java
/*
 * Copyright 2020 ThoughtWorks, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.thoughtworks.go.domain;

import com.thoughtworks.go.config.CaseInsensitiveString;
import com.thoughtworks.go.config.CruiseConfig;
import com.thoughtworks.go.config.materials.Materials;
import com.thoughtworks.go.domain.buildcause.BuildCause;
import com.thoughtworks.go.domain.label.PipelineLabel;

import java.io.File;
import java.util.Date;
import java.util.Map;

public class Pipeline extends PersistentObject implements PipelineInfo {

    private String pipelineName;
    private Integer counter;
    private PipelineLabel pipelineLabel = PipelineLabel.defaultLabel();
    private Stages stages;
    private BuildCause buildCause;
    private double naturalOrder;

    public Pipeline() {
        this(new Stages());
    }

    private Pipeline(Stages stages) {
        this.stages = stages;
        this.counter = 0;//counter can never be null, as pipeline identifier creation with null counter is not allowed
    }

    public Pipeline(String pipelineName, String labelTemplate, BuildCause buildCause, EnvironmentVariables envVars, Stage... stages) {
        this(new Stages(stages));
        this.pipelineName = pipelineName;
        this.buildCause = buildCause;
        this.pipelineLabel = PipelineLabel.create(labelTemplate, envVars);
    }

    public Pipeline(String pipelineName, BuildCause buildCause, Stage... stages) {
        this(pipelineName, PipelineLabel.COUNT_TEMPLATE, buildCause, new EnvironmentVariables(), stages);
    }

    public File defaultWorkingFolder() {
        return new File(CruiseConfig.WORKING_BASE_DIR + pipelineName);
    }

    @Override
    public String getName() {
        return pipelineName;
    }

    public void setName(String pipelineName) {
        this.pipelineName = pipelineName;
    }

    public void setCounter(Integer counter) {
        this.counter = counter;
    }

    public Integer getCounter() {
        return counter;
    }

    public Stages getStages() {
        return stages;
    }

    public void setStages(Stages stages) {
        this.stages = stages;
    }

    public Stage getFirstStage() {
        return stages.first() == null ? new NullStage("unknown") : stages.first();
    }

    public Materials getMaterials() {
        return buildCause.materials();
    }

    public BuildCause getBuildCause() {
        return buildCause;
    }

    public void setBuildCause(BuildCause buildCause) {
        this.buildCause = buildCause;
    }

    public void setModificationsOnBuildCause(MaterialRevisions materialRevisions) {
        this.buildCause.setMaterialRevisions(materialRevisions);
    }

    public String getBuildCauseMessage() {
        if (buildCause == null) {
            return "Unknown";
        }
        //TODO: ChrisS: Replace this with encapsulation and null checks for the buildCause
        //also, we may choos to persist the real username and buildcause rather than throwing
        // away that information as we do now
        String message = "Unknown";
        if (getBuildCause() != null && getBuildCause().getBuildCauseMessage() != null) {
            message = getBuildCause().getBuildCauseMessage();
        }
        return message;
    }

    public ModificationSummaries toModificationSummaries() {
        if (buildCause == null) {
            return new ModificationSummaries();
        } else {
            return buildCause.toModificationSummaries();
        }
    }

    @Override
    public boolean hasStageBeenRun(String stageName) {
        return stages.hasStage(stageName);
    }

    public String getLabel() {
        return pipelineLabel.toString();
    }

    /**
     * Will not apply template logic; to be used only in testing or by iBatis
     *
     * @param label
     */
    public void setLabel(String label) {
        this.pipelineLabel.setLabel(label);
    }

    public void updateCounter(Integer lastCount) {
        counter = lastCount + 1;
        updateLabel();
    }

    private void updateLabel() {
        Map<CaseInsensitiveString, String> namedRevisions = this.getMaterialRevisions().getNamedRevisions();
        this.pipelineLabel.updateLabel(namedRevisions, counter);
    }

    @Override
    public boolean isAnyStageActive() {
        return stages.isAnyStageActive();
    }

    @Override
    public String nextStageName(String stageName) {
        return stages.nextStageName(stageName);
    }

    public Stage findStage(String stageName) {
        return stages.byName(stageName);
    }

    // used by ibatis
    public void setBuildCauseMessage(String buildCauseMessage) {
        buildCause.setMessage(buildCauseMessage);
    }

    // used by ibatis
    public void setApprovedBy(String approvedBy) {
        buildCause.setApprover(approvedBy);
    }

    public MaterialRevisions getMaterialRevisions() {
        return buildCause.getMaterialRevisions();
    }

    public Date getModifiedDate() {
        return buildCause.getModifiedDate();
    }

    public PipelineIdentifier getIdentifier() {
        return new PipelineIdentifier(pipelineName, counter, getLabel());
    }

    public double getNaturalOrder() {
        return naturalOrder;
    }

    public void setNaturalOrder(double naturalOrder) {
        this.naturalOrder = naturalOrder;
    }

    public EnvironmentVariables scheduleTimeVariables() {
        return buildCause.getVariables();
    }

    public boolean isBisect() {
        return naturalOrderIsNotAnInteger();
    }

    private boolean naturalOrderIsNotAnInteger() {
        return naturalOrder - new Double(naturalOrder).intValue() > 0;
    }
}
